s390x: pattern match saturated truncation

folkertdev opened the pull request:
fixes https://github.com/llvm/llvm-project/issues/153655

This does not yet work, but should make it easier to talk about.

After legalizing `smin` and `smax` (I'm still matching the older patterns below, not sure if that is really needed any more?), the DAG is 

```
Optimized legalized selection DAG: %bb.0 'i32_signed:bb2'
SelectionDAG has 17 nodes:
  t0: ch,glue = EntryToken
  t17: v4i32 = BUILD_VECTOR Constant:i32<-32768>, Constant:i32<-32768>, Constant:i32<-32768>, Constant:i32<-32768>
  t16: v4i32 = BUILD_VECTOR Constant:i32<32767>, Constant:i32<32767>, Constant:i32<32767>, Constant:i32<32767>
          t2: v4i32,ch = CopyFromReg t0, Register:v4i32 %0
        t18: v4i32 = smax t2, t17
      t20: v4i32 = smin t18, t16
          t4: v4i32,ch = CopyFromReg t0, Register:v4i32 %1
        t19: v4i32 = smax t4, t17
      t21: v4i32 = smin t19, t16
    t50: v8i16 = SystemZISD::PACK t20, t21
  t14: ch,glue = CopyToReg t0, Register:v8i16 $v24, t50
  t15: ch = SystemZISD::RET_GLUE t14, Register:v8i16 $v24, t14:1
```

However, the pattern still does not match. I'm now suspicious of those `BUILD_VECTOR`s, because the rest just looks like it should work.

I'm also aware that the order of the patterns matters, which is why the `VPKSF` pattern is now before `VPKF`.

cc @uweigand

folkertdev committed:
s390x: legalize smin/smax/umin/umax for vectors

folkertdev committed:
s390x: map saturating truncation to a packs/packu

folkertdev commented on llvm/lib/Target/SystemZ/SystemZInstrVector.td:
```diff
@@ -1269,11 +1282,24 @@ multiclass IntegerMinMaxVectorOps<ValueType type, SDPatternOperator cmph,
+  defm : IntegerMinMaxVectorOps<type, cmph, min, max>;
+
+  let Predicates = [FeatureVector] in {
+    def : Pat<(type (smax VR128:$x, VR128:$y)),
```
How would I verify that? just remove `IntegerMinMaxVectorOps` and run the tests, see if they still succeed?

uweigand commented on llvm/lib/Target/SystemZ/SystemZOperators.td:
```diff
@@ -1067,6 +1067,15 @@ def vsplat_imm_eq_1 : PatFrag<(ops), (build_vector), [{
 
+def vsplat_imm_eq_s32_min : PatFrag<(ops), (build_vector), [{
+  APInt Imm;
+  return ISD::isConstantSplatVector(N, Imm) && Imm.getSExtValue() == -32768;
```
Ah, right.  You can probably generalize this function to determine the type (from `N`) and check for the type's minimum/maximum values programmatically.

uweigand commented on llvm/lib/Target/SystemZ/SystemZInstrVector.td:
```diff
@@ -680,39 +696,39 @@ let Predicates = [FeatureVector] in {
+    def VMXF : BinaryVRRc<"vmxf", 0xE7FF, smax, v128f, v128f, 2>;
+    def VMXG : BinaryVRRc<"vmxg", 0xE7FF, smax, v128g, v128g, 3>;
     let Predicates = [FeatureVectorEnhancements3] in
       def VMXQ : BinaryVRRc<"vmxq", 0xE7FF, null_frag, v128q, v128q, 4>;
```
We might want to make min/max legal for i128 as well (when we have FeatureVectorEnhancements3).

folkertdev replied:
Done, it causes some trivial register renames in the tests, but otherwise seems to work ok.

uweigand reviewed (commented):
Looking pretty good now!   I'm wondering - should we handle PCKLS similarly?

folkertdev committed:
s390x: legalize smin/smax/umin/umax for v128q

uweigand reviewed (commented):
Looks all good to me.  Only remaining question is whether we can drop the `IntegerMinMaxVectorOps` patterns now - if the tests still all pass, we should just do it as part of this patch.

folkertdev committed:
s390x: remove `IntegerMinMaxVectorOps` multiclass

The more general smin/umin/smax/umax are now legal, so these patterns are no longer needed

folkertdev commented:
Looks like that works out, `bin/llvm-lit ../llvm/test/CodeGen/SystemZ` is clean locally, so I've removed those patterns.

uweigand reviewed (commented):
Excellent, thanks!  I think this is ready to move out of draft status.

uweigand reviewed (approved):
LGTM assuming CI passes.  Thanks for working on this!

folkertdev commented:
Thank you for guiding me through this! LLVM can be kind of inscrutable, so that help is really useful.

Am I supposed to do something to merge this?

uweigand committed:
Merge branch 'main' into s390x-saturated-truncation
